language/diem-framework/modules/Vote.move

```diff
@@ -0,0 +1,377 @@
+    /// after this vote and false represents that the ballot has not been
+    /// approved after this vote
+    public fun vote<Proposal: store + drop>(
+        voter_account: &signer,
+        ballot_id: BallotID,
+        proposal_type: vector<u8>,
+        proposal: Proposal,
+    ): bool acquires Ballots {
+        let ballot_data = borrow_global_mut<Ballots<Proposal>>(ballot_id.proposer);
+
+        // Remove any expired ballots
+        gc_internal<Proposal>(ballot_data);
+
+        let ballots = &mut ballot_data.ballots;
+        let i = 0;
+        let len = Vector::length(ballots);
+        while (i < len) {
+            if (&Vector::borrow(ballots, i).ballot_id == &ballot_id) break;
+            i = i + 1;
+        };
+        assert(i < len, Errors::invalid_state(EBALLOT_NOT_FOUND));
+        let ballot_index = i;
+        let ballot = Vector::borrow_mut(ballots, ballot_index);
+
+        assert(&ballot.proposal == &proposal, Errors::invalid_argument(EBALLOT_PROPOSAL_MISMATCH));
+        assert(&ballot.proposal_type == &proposal_type, Errors::invalid_argument(EBALLOT_PROPOSAL_MISMATCH));
+
+        let voter_address = Signer::address_of(voter_account);
+        let voter_address_bcs = BCS::to_bytes(&voter_address);
+        let allowed_voters = &ballot.allowed_voters;
+
+        assert(check_voter_present(allowed_voters, &voter_address_bcs), Errors::invalid_state(EINVALID_VOTER));
+        assert(DiemTimestamp::now_seconds() <= ballot.expiration_timestamp_secs, Errors::invalid_state(EBALLOT_EXPIRED));
+
+        assert(!check_voter_present(&ballot.votes_received, &voter_address_bcs), Errors::invalid_state(EALREADY_VOTED));
+
+        let i = 0;
+        let len = Vector::length(allowed_voters);
+        while (i < len) {
+            let weighted_voter = Vector::borrow(allowed_voters, i);
+            if (&weighted_voter.voter == &voter_address_bcs) {
+                Vector::push_back(&mut ballot.votes_received, *(weighted_voter));
+                ballot.total_weighted_votes_received = ballot.total_weighted_votes_received + weighted_voter.weight;
+                Event::emit_event<VotedEvent>(
+                    &mut ballot_data.voted_handle,
+                    VotedEvent {
+                        ballot_id: *(&ballot_id),
+                        voter: voter_address,
+                        vote_weight: weighted_voter.weight,
+                    },
+                );
+                break
+            };
+            i = i + 1;
+        };
+        let ballot_approved = ballot.total_weighted_votes_received >= ballot.num_votes_required;
+        // If the ballot gets approved, remove the ballot immediately
+        if (ballot_approved) {
+            Vector::swap_remove(ballots, ballot_index);
+            Event::emit_event<BallotApprovedEvent>(
+                &mut ballot_data.ballot_approved_handle,
+                BallotApprovedEvent {
+                    ballot_id,
+                },
+            );
+        };
```

Comment:
I'm not sure I follow what is supposed to happen after this. If the vote causes the ballot to be approved what proof is there that this ballot has been approved on chain (and that the approval hasn't already been used). i.e., how does the action that is being voted on guarantee that:
1. The action it is about to perform has been voted on and accepted; and
2. That this action has not already been performed

Reply:
I'm open to suggestions. 
> If the vote causes the ballot to be approved what proof is there that this ballot has been approved on chain
The only proof that we have is the event that is emitted (`BallotApprovedEvent`).

> Is this only the case if this returns true and then in that same transaction it will have to perform the action?
Yes, that was my current assumption. 

Let's discuss this tomorrow - there are many ways of going about doing this each with certain tradeoffs.

Reply:
I was thinking about this this morning a bit more, and I actually quite like this now that I'm understanding what's going on and how this would fit in with the calling code. I think it's actually a a very nice clean way of doing it.